feat(RELEASE-1675): auto generate README.md files for tasks/pipelines (#1)

Paul123111 opened the pull request:
## Describe your changes
Add readme_generator.sh, which generates README.md files for tasks and pipelines, as well as a check_readme.sh script and workflow to check that the README.md files in each task and pipeline match the output of readme-generator.sh.

## Relevant Jira
https://issues.redhat.com/browse/RELEASE-1675

## Checklist before requesting a review
- [x] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)

FilipNikolovski commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+  echo "  --help: Show this help message"
+  echo
+  echo "Items are task or pipeline directories. They can be supplied"
+  echo "either as arguments or via the README_ITEMS environment variable."
```
Just curious, is there a reason why the script supports both an env var and args?

Paul123111 replied:
Mainly because `test_tekton_tasks.sh` also supports both, but also to make it more convenient to run locally.

FilipNikolovski replied:
cool, thanks for clarifying.

FilipNikolovski commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+      show_help
+      ;;
+    *)
+      CLI_README_ITEMS+="$1 "
```
You could make CLI_README_ITEMS an array, to avoid potential issues like if a directory contains a space or something.
```suggestion
      CLI_README_ITEMS+=("$1")
```

Paul123111 replied:
fixed

seanconroy2021 commented on .github/workflows/check_readme.yaml:
```diff
@@ -0,0 +1,31 @@
+---
+name: Check README.md
```
Could we not just add this to the existing linting workflow? Why does it need to be separate?
https://github.com/konflux-ci/release-service-catalog/blob/f7724b5c7d75ade7500dde0ebb738179ace03fe6/.github/workflows/lint.yaml#L51

Paul123111 replied:
fixed

seanconroy2021 commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+    exit 1
+  fi
+
+  if [[ "$(hack/readme_generator.sh --dry-run "$ITEM_DIR")" != "$(cat "$README_PATH")" ]]; then
```
This manual string comparison, followed by a separate diff, basically duplicates work. Why can't we use diff in the expression and use the exit code?

Paul123111 replied:
fixed - adding diff output to variable, then checking exit code

seanconroy2021 replied:
`if ! diff -u <(hack/readme_generator.sh --dry-run "$ITEM_DIR") "$README_PATH"; then`

You could just simplify it to this and drop the extra echo (-u will print the mismatch). You don't need to store it as a variable, you can use the diff as the expression and invert it.

Paul123111 replied:
Fixed, thanks!

seanconroy2021 commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+    diff <(hack/readme_generator.sh --dry-run "$ITEM_DIR") <(cat "$README_PATH")
+    echo "  Error: README.md has not been updated. Please use hack/readme-generator.sh to" \
+      "generate a new README.md to replace $ITEM/README.md"
+    exit 1
```
We probably shouldn't exit on the first mismatch, as if users have multiple readme changed, we will want to see all errors and then exit

Paul123111 replied:
fixed - now exits at the end, lists failed tasks/pipelines

FilipNikolovski commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+if [[ -n "${CLI_README_ITEMS[@]}" ]]; then
+  README_ITEMS="${CLI_README_ITEMS[@]}"
+else
+  README_ITEMS="${README_ITEMS:-}"      # Use env var or empty string
```
You'll probably want to read the env var as an array here as well, something like:
```
read -r -a README_ITEMS <<< "$README_ITEMS"
```
And then use it in the loop like:
```
for ITEM in "${README_ITEMS[@]}"; do
...
```

Paul123111 replied:
fixed

FilipNikolovski commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+  README_ITEMS="${README_ITEMS:-}"      # Use env var or empty string
+fi
+
+if [ -z "${README_ITEMS}" ]
```
```suggestion
if [ "${#README_ITEMS[@]}" -eq 0 ]
```

Paul123111 replied:
fixed

FilipNikolovski commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,102 @@
+  esac
+done
+
+if [[ -n "${CLI_README_ITEMS[@]}" ]]; then
```
```suggestion
if [[ "${#CLI_README_ITEMS[@]}" -gt 0 ]]; then
```

Paul123111 replied:
fixed

Paul123111 commented:
Added `check_param_punctuation.sh` and workflow from Martin's comment in one of the other PRs.

seanconroy2021 commented on .github/workflows/lint.yaml:
```diff
@@ -109,3 +109,49 @@ jobs:
+        env:
+          README_ITEMS: >-
+            ${{ steps.changed-dirs.outputs.all_changed_files }}
+  check-param-punctuation:
```
Why do we need another job? Can we not do it all in `check README.md files`

Paul123111 replied:
Fixed - `check param punctuation` is now a step in `check README.md files` job

seanconroy2021 replied:
Could we not keep it all one script? It would become annoying to maintain two

Paul123111 replied:
Fixed - `check_readme.sh` does both now

seanconroy2021 commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,141 @@
+do
+  echo "Task/Pipeline item: $ITEM"
+  ITEM_NAME=$(basename "$ITEM")
+  ITEM_DIR=$(echo "$ITEM" | cut -d '/' -f -3)
```
Can we not do a here string (<<<) instead of echo

Paul123111 replied:
fixed

seanconroy2021 commented on .github/scripts/check_readme.sh:
```diff
@@ -0,0 +1,141 @@
+
+CLI_README_ITEMS=()
+FAILED_PARAMS=()
+
```
You haven't declared FAILED_ITEMS=()

Paul123111 replied:
fixed

seanconroy2021 commented on .github/workflows/lint.yaml:
```diff
@@ -109,3 +109,25 @@ jobs:
+        run: .github/scripts/check_readme.sh
+        env:
+          README_ITEMS: >-
+            ${{ steps.changed-dirs.outputs.all_changed_files }}
```
Should this not be changed dirs

Paul123111 replied:
Are you sure? I can't find any `all_changed_dirs` option in https://github.com/tj-actions/changed-files/blob/main/README.md#outputs-, and every other workflow in `lint.yaml` also uses `all_changed_files`.

Paul123111 replied:
I think `dir_names: true` makes so that only directories are selected

seanconroy2021 replied:
Yeah, you're correct, I was. looking at the id

seanconroy2021 commented on hack/readme_generator.sh:
Why is this in hack? Can we not move it into scripts?

Paul123111 replied:
I think I remember Johnny saying something about putting the README generator script into the `hack` directory in one of the stand-ups, which is why I put there. I think it could also in scripts, but I'll ask Johnny about which directory it should be in when he's back.

Paul123111 replied:
Fixed - David said to put the script in `.github/scripts` - I've removed `hack` and moved the script

johnbieren reviewed (commented):
The error messages in the script look good, but can you also please add a note in CONTRIBUTING.md about the script? Under the submitting changes section? Bonus points if you add a bullet point to pull_request_template.md like `I have run the README.md generator script in .github/scripts/readme_generator.sh`

Commit:
feat(RELEASE-1675): auto generate README.md files for tasks/pipelines

Add readme_generator.sh, which generates README.md files
for tasks and pipelines, as well as a check_readme.sh script and
workflow to check that the README.md files in each task and pipeline
match the output of readme-generator.sh. Also updated CONTRIBUTING.md
to mention this new requirement.

Add check in check_readme.sh to stop '.' from being
added to the end of task/pipeline param descriptions.

Signed-off-by: Paul Fitzgerald <pafitzge@redhat.com>

johnbieren commented on CONTRIBUTING.md:
```diff
@@ -125,6 +125,31 @@ Here is an example
+### Keeping Documentation Up to Date
+
+Whenever a task or pipeline is changed, please run the `.github/scripts/readme_generator.sh` script with the
+changed task/pipeline directories as arguments to update the README.md description and table included with the
```
`table included with the table` is confusing to me. Maybe just `README.md description and parameter table`?

Paul123111 replied:
Fixed - `table included with the table` was a mistake, it was just supposed to be `table`
